src/connection.rs

```diff
@@ -162,6 +166,45 @@ impl<T: Read + Write + ScmSocket> HttpConnection<T> {
             .ok_or(ConnectionError::ParseError(RequestError::Overflow))
     }
 
+    /// Receive data along with optional files descriptors.
+    /// It is a wrapper around the same function from vmm-sys-util.
+    ///
+    /// # Errors
+    /// `StreamError` is returned if any error occurred while reading the stream.
+    fn recv_with_fds(&mut self) -> Result<(usize, Option<Vec<File>>), ConnectionError> {
+        let buf = &mut self.buffer[self.read_cursor..];
+        // We must allocate the maximum number of receivable file descriptors
+        // if don't want to miss any of them. Allocating a too small number
+        // would lead to the incapacity of receiving the file descriptors.
+        let mut fds = [0; SCM_MAX_FD];
+        let mut iovecs = [libc::iovec {
+            iov_base: buf.as_mut_ptr() as *mut libc::c_void,
+            iov_len: buf.len(),
+        }];
+
+        // Safe because we have mutably borrowed buf and it's safe to write
+        // arbitrary data to a slice.
+        let (read_count, fd_count) = unsafe {
+            self.stream
+                .recv_with_fds(&mut iovecs[..], &mut fds)
+                .map_err(ConnectionError::StreamReadError)?
+        };
+
+        let files = if fd_count == 0 {
+            None
+        } else {
+            let mut files = Vec::new();
+            for fd in fds.iter().take(fd_count) {
+                // Safe because all fds are owned by us after they have been
+                // received through the socket.
+                files.push(unsafe { File::from_raw_fd(*fd) });
+            }
+            Some(files)
+        };
+
+        Ok((read_count, files))
+    }
```

Comment:
Let's not reimplement this in `micro-http`. You can use [the one](https://github.com/rust-vmm/vmm-sys-util/blob/b4c071252f1f10100e3cee956e145f8dc3b31845/src/linux/sock_ctrl_msg.rs#L405) in `vmm-sys-util` [1] while we still wait for the std implementation to become stable.

[1] - https://github.com/rust-vmm/vmm-sys-util/blob/b4c071252f1f10100e3cee956e145f8dc3b31845/src/linux/sock_ctrl_msg.rs#L405

Reply:
But I'm actually relying on the one from `vmm-sys-util`, it's just that I need to prepare things from `micro-http`. Does that make sense?

Reply:
Right, my mistake, I didn't even look at it right and just assumed it was re-implemented :see_no_evil: